[inductor][triton] improved kernel-level benchmarking

nmacchioni opened the pull request:
TLDR; by improving our kernel-level benchmarking we can reduce compile time overhead by ~12% for training default on torchbench, ~25% for inference default on torchbench, while maintaining near same or higher performance.

1.  Write our own custom benchmarking method
2. Move everything benchmarking out of torch._inductor.runtime.runtime_utils and into torch._inductor.runtime.benchmarking

Some things that are very different from the original implementation:
- We now try to torch.cuda._sleep through the CPU launch overhead during our benchmarking iteration. This gives us better grouping on the GPU side and can reduce some noise in our output. We use the estimation loop to measure the CPU launch overhead of a single iteration and deduce the total CPU launch overhead from that measurement. However, because of all this we now run the risk of hitting the GPU's max queue limit. We workaround this by experimentally deciding the max queue limit, then breaking the benchmarking into chunks that fix within the limit.
- Benchmark ranking: an early termination of the benchmarker (immediately after the estimation loop) which has experimentally shown to significantly reduce compile time overhead without affecting performance. Essentially, when we benchmark a groups of kernels where there times are not important but their ranking is (template selection) this can be beneficial.
- Benchmark pruning: prune away noticeably bad kernels after the estimation loop. Just like benchmark ranking, this reduces compile time overhead by taking at most the top X (default 5) kernels that have a runtime less than or equal to 
the minimum runtime * some factor (default 1.1). The un-pruned kernels will proceed through the entire benchmarking loop.
- Lazy benchmark: lazily evaluated benchmarks, essentially just a way to get ranking and pruning working. If we can lazily evaluate benchmarks, we can actually group together like benchmarks (without user having to change code). Example is GEMM autotuning, ~22 different choices if we lazy benchmark each of them then we will only run a single benchmark_many_gpu which will do all 22 choices at once.

**Training Results**
<img width="1728" alt="Screenshot 2024-07-23 at 10 44 06 AM" src="https://github.com/user-attachments/assets/2c3f796b-7a87-4bea-9402-9d809930ba19">
<img width="1706" alt="Screenshot 2024-07-23 at 10 45 00 AM" src="https://github.com/user-attachments/assets/dbdf4953-8b02-43a7-8555-ce780aa6d1ec">

**Inference Results**
<img width="1710" alt="Screenshot 2024-07-23 at 10 44 29 AM" src="https://github.com/user-attachments/assets/f1b7ec66-8438-4a48-8d26-6a0184bdca73">
<img width="1710" alt="Screenshot 2024-07-23 at 10 44 42 AM" src="https://github.com/user-attachments/assets/06be7f00-0a09-4ad8-bd41-9188aec65df5">


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec @peterbell10

nmacchioni committed:
initial commit

nmacchioni committed:
fix triton wheel build

nmacchioni committed:
exclude benchmarking from single record tests

nmacchioni committed:
temporarily bypass kernel verify

nmacchioni committed:
fix triton wheel build for rocm

nmacchioni committed:
fix LazyBenchmark making its way into test_max_autotune

nmacchioni committed:
refactor benchmarking (@cached_properties and counters)

nmacchioni committed:
add basic testing

nmacchioni committed:
more testing + counter fixes

nmacchioni committed:
fix improper use of cuda events

nmacchioni committed:
small fixes

nmacchioni committed:
fix parts of benchmark iters and sleep cycles calculations

nmacchioni committed:
major refactor, more comments, initialize callables before queueing in lazy benchmarking, ...

nmacchioni committed:
second attempt at getting testing to run

nmacchioni committed:
fix rounding error causing some cuda events to not be recorded

nmacchioni committed:
temporarily record history for cudagraphs

nmacchioni committed:
flake8

nmacchioni committed:
more lints and type fixes

nmacchioni committed:
temporary log output code

nmacchioni committed:
Revert "temporary log output code"

This reverts commit 40e8c843aa822c1feb3712436900d2de475ff45d.

nmacchioni committed:
temporary log output code

nmacchioni committed:
fix more type checking errors

nmacchioni committed:
Revert "temporary log output code"

This reverts commit 86761b124c320f772e2c394af26098a365ecb168.

nmacchioni committed:
temporarily enable benchmarking and output_code logs

nmacchioni committed:
assert unreachable is not in typing

nmacchioni committed:
hunting down memory leaks

nmacchioni committed:
Revert "temporarily enable benchmarking and output_code logs"

This reverts commit 554940ef6430b292a0798033b5002e2297ee345d.

nmacchioni committed:
Revert "hunting down memory leaks"

This reverts commit 36f983c899e49d8f52f91aea4da08a9c690b4ef6.

nmacchioni committed:
temporarily disable lazy benchmarking outside of triton

nmacchioni committed:
Revert "temporarily record history for cudagraphs"

This reverts commit 346f99c5baa59ce96e476426cdd39e5c9eec35fc.

nmacchioni committed:
update lazy benchmark magic functions

nmacchioni committed:
readd lazy benchmarking in triton heuristics with changes

nmacchioni committed:
typing + small fixes

nmacchioni committed:
don't use static method for magic function wrapper

nmacchioni committed:
go back to regularly defined magic methods

nmacchioni committed:
hunt down memory leak

nmacchioni committed:
hunting leak in lazy benchmark

nmacchioni committed:
hunting leak in lazy benchmark 2/x

nmacchioni committed:
hunting leak in lazy benchmark 3/x

nmacchioni committed:
hunting leak in lazy benchmark 4/x

nmacchioni committed:
renable lazy benchmarking

nmacchioni committed:
add comments explaining fix 1/x and 2/x for memory leak

nmacchioni committed:
fix b2b gemm test

nmacchioni committed:
lint 1/x

nmacchioni committed:
lint 2/x

nmacchioni committed:
lint 3/x

nmacchioni committed:
lint 4/x

nmacchioni committed:
lint 5/x

nmacchioni committed:
lint 6/x

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

bump margin of error for cpu benchmark validity, make pass since this is a clone of the original function

nmacchioni committed:
Update benchmarking.py

refactor some methods to cached properties

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
lint 7/x

nmacchioni committed:
small fixes

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
remove unneeded counters

nmacchioni committed:
sleep 1ms before estimation loop in benchmark_many_gpu

nmacchioni committed:
pruning limit, lower default pruning_factor

nmacchioni committed:
working towards landable version (JK for falling back to original implementation i.e. do_bench, and JKs for enabling lazy benchmarking, ranking, and pruning)

nmacchioni committed:
fix typing

nmacchioni committed:
fix typing again

nmacchioni committed:
fix dropped self

nmacchioni committed:
fix dropped self

nmacchioni committed:
fix missing kwargs

nmacchioni committed:
refactor test_benchmarking

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
readd refactored tests

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Merge branch 'main' into nmacchioni-benchmarking

nmacchioni committed:
lints + fixes + remove unrelated lints

Update build_triton_wheel.py

Update tensor_layout_mini_benchmark.py

Update tensor_layout_mini_benchmark.py

Update test_b2b_gemm.py

Update test_inductor_utils.py

Update test_padding.py

Update test_scatter_optimization.py

Update select_algorithm.py

Update select_algorithm.py

Update select_algorithm.py

Update scheduler.py

Update triton_heuristics.py

Update ir.py

Update triton.py

nmacchioni committed:
subprocess autotune doesn't work well with LazyBenchmarks

nmacchioni committed:
Update config.py

nmacchioni committed:
small changes

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
nits

bertmaher commented on .ci/docker/common/install_triton.sh:
```diff
@@ -19,7 +19,7 @@ elif [ -n "${XPU_VERSION}" ]; then
   TRITON_TEXT_FILE="triton-xpu"
 else
-  TRITON_REPO="https://github.com/openai/triton"
+  TRITON_REPO="https://github.com/nmacchioni/triton"
```
What Triton changes are needed to support this diff?  Can you link a PR or something in the summary?

nmacchioni replied:
Here is probably the best place to brainstorm how we could upstream these changes to Triton, I figured our best chance would be adding a switch to Triton's do_bench that falls back to our benchmarker instead.

nmacchioni replied:
Once we're approved here we can work with OpenAI to make this accessible from Triton, stack whatever that change ends up being on top of the release branch and bump the Triton commit hash by one to that PR. This should be safe in OSS (assuming all tests are passing, and after I do a final performance run), and internally (everything is gated by JKs and the current default is to fallback to the original benchmarking method), but of course we will test there as well.

bertmaher commented on test/inductor/test_padding.py:
```diff
@@ -213,9 +215,13 @@ def f(**inputs):
 
         print("benchmark for good shape")
-        latency_good_shape = do_bench(lambda: f_good_shape(**inputs_good_shape))
+        latency_good_shape = benchmarker.benchmark(
```
Kind of a pervasive comment - this diff changes `do_bench` to `benchmarker.benchmark`, which means there are a bunch of mechanical changes to slog through without an actual API change.  Can we keep `do_bench` and just internally have it call `benchmarker.benchmark`?

nmacchioni replied:
Sure, I'll add this for now (but might clean it up in the future)

albanD commented on torch/_C/__init__.pyi.in:
```diff
@@ -1946,6 +1946,7 @@ class _CudaDeviceProperties:
     gcnArchName: str
     warp_size: _int
     uuid: str
+    L2_cache_size: _int
```
This one follows the other properties on this cuda struct and is pretty small so the change here sounds good, no need to split it up. FYI @eqy

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
do_bench dispatch

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1026 @@
+benchmarker = Benchmarker()
+
+
+def do_bench(callable_or_callables: Union[Callable[..., Any], List[Callable[..., Any]]], *args: Any, lazy: bool = False, **kwargs: Any) -> Union[LazyBenchmark, float]:
```
@bertmaher Added this do_bench wrapper, its not perfect (mainly because we have no way of differentiating CPU and GPU callables once they've been wrapped to be callable by the benchmarker) but it does reduce nominal changes.

nmacchioni replied:
Eh actually I don't like how this silently dispatches to GPU since it can't differentiate the devices, I'm going to revert the change. It does result in more nominal changes like you mentioned, but I think it is worth having the clarity of knowing what is called.

nmacchioni committed:
type fixes

nmacchioni committed:
revert do_bench wrapper

eellison commented on benchmarks/dynamo/microbenchmarks/tensor_layout_mini_benchmark.py:
```diff
@@ -54,8 +54,8 @@ def test_fn():
 
-    baseline_ms = do_bench(baseline_fn, rep=40)
-    test_ms = do_bench(test_fn, rep=40)
+    baseline_ms = benchmarker.benchmark(baseline_fn, (), {}, rep=40)
```
Would be great if we could migrate callsites as needed instead of changing the api - necessitating args and kwargs in every call is an ergonomic downgrade

nmacchioni replied:
Tbh I think this is already either not working or I'm missing something super obvious, because torch._inductor.runtime.runtime_utils.do_bench does already necessitate args and kwargs

nmacchioni replied:
readded do_bench to runtime_utils, just redirects to benchmarker

eellison commented on test/inductor/test_max_autotune.py:
```diff
@@ -52,7 +52,10 @@ def _get_path_without_sccache() -> str:
+    # choice.benchmark outputs a LazyBenchmark is some cases,
+    # we need to initialize the lazy value by converting it
+    # to a float
+    result = float(choice.benchmark(*args, out=out))
```
similarly, i dont think we should change the output type by default.

nmacchioni replied:
Fixed

eellison reviewed (commented):
Very cool wins and great work ! I haven't done a thorough review yet of all the code. Could we put BC incompatible api changes in a separate pr, or explicitly opt into laziness? I don't know that laziness a good idea by default. Lazily evaluating could lead to memory increase. There only a few places in the codebase that are really invoked enough where we would expect noticeable wins by adopting laziness.

Also, this a large amount of code, and because the changes are not split up it's hard to say which changes are responsible for speedups and which changes might not necessarily be worth the complexity.

nmacchioni committed:
no bc changes 1/x

nmacchioni committed:
no bc changes 2/x

nmacchioni committed:
no bc changes 3/x

nmacchioni committed:
no bc changes 4/x

nmacchioni committed:
no bc changes 5/x

nmacchioni committed:
no bc changes 6/x

nmacchioni committed:
no bc changes 7/x

nmacchioni commented:
> Could we put BC incompatible api changes in a separate pr, or explicitly opt into laziness?

Removed BC changes, and added explicit opt into laziness where applicable. Also, removed some lazy callsites that were easily refactorable into direct grouped benchmarking (i.e. benchmark_many_gpu).


> I don't know that laziness a good idea by default. Lazily evaluating memory could lead to memory increase. There only a few places in the codebase that are really invoked enough where we would expect noticeable wins by adopting laziness.

Removed a couple of the lazy callsites that were 100% needed, with small changes. The areas where we really want laziness (GEMM autotuning and other generic template autotuning) should not increase memory usage (and I believe this is confirmed by the performance runs) since all of the input/output tensors are already shared for each individual template and are allocated for the same amount of time. I personally like the laziness in the cases I mentioned, but we can definitely go through in followup PRs to refactor the callsites and explicitly benchmark multiple templates at once.

> Also, this a large amount of code, and because the changes are not split up it's hard to say which changes are responsible for speedups and which changes might not necessarily be worth the complexity.

I should have better documented the (enormous amount) of performance runs I did while writing this, but it sort of breaks down like this:

Performance wins
- ~100% attributable to the more accurate benchmarking, which really did require all the weirdness (GPU sleeps, memory warmups, grouped benchmarking, etc.), and I believe most of these (small) wins come from the more accurate fusions

Compile time wins
- Excluding pruning/ranking/grouped benchmarking we are ~1.5x faster than the original benchmarking function for default PT2 (~25ms per benchmark vs ~40ms)
- Grouped benchmarking also doesn't help much with compile time, but it shaves off ~2ms for every function after the first in a group (so 10 benchmarks is ~230ms not 250ms)
- Pruning helps a lot for max-autotune mode, since we go from ~20 to 5 configs at max, so ~500ms -> ~125ms
- Ranking helps the most overall imo, since it takes generic template autotuning down from ~40ms per config to basically nothing (since we only do 5 iterations per callable).

nmacchioni committed:
Update benchmarking.py

nmacchioni committed:
undo triton pin changes

nmacchioni committed:
types

nmacchioni committed:
nits

nmacchioni committed:
nits

nmacchioni commented:
@eellison @bertmaher I went ahead and removed the pin change for Triton, it wouldn't have been landable in that form anyways so let's see where we're at with perf and compile time minus that.

nmacchioni committed:
Merge commit '8927fc209f4ce73fd642fd1e9cc4264a07921c8f' into nmacchioni-benchmarking

nmacchioni committed:
lints, types, nits

nmacchioni committed:
fix failing tests

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
fix counter spelling

nmacchioni committed:
temp renable triton changes for benchmarking

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
temp undo shuffle

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
enable more lazy benchmarks

nmacchioni committed:
fix logging

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Revert "temp renable triton changes for benchmarking"

This reverts commit 4d2146421d37c89655aaa4718f730b6ce6eba7f2.

xuzhao9 commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+    )
+
+
+class Benchmarker:
```
Do you think it make sense to have both an "online" and an "offline" benchmarker that works in different scenarios? For example, the offline benchmarker has a more relaxed constraint of benchmarking time (like using cuda graph)

nmacchioni replied:
My guess is that using cudagraphs to benchmark wouldn't actually help with performance enough to warrant the complexity and reliability concerns that would come from trying to offline the benchmarking, if it even helped with performance at all.

The one concrete thing we get from using cudagraphs to benchmark is the ability to exclude GPU launch overhead from our measurements, but I'm not convinced this matters is 99% of cases. The most basic use of a benchmarker is autotuning a kernel to one set of optimal configs, for which GPU launch overhead is basically irrelevant since it is fairly stable and should affect each config in an equal and linear fashion (so we still get the same ordering of configs, which is all that is important). The singular case I can think of when GPU launch overhead might play a factor in actual E2E performance is when we are comparing fusion opportunities. If we want to try fusing some kernels A + B into a new C kernel we first have to check that the runtime of C is less than the runtime of A + B to know if it is a "good" fusion, the problem here is that we are measuring the GPU launch overhead twice when benchmarking A + B and only once when benchmarking C. But then again, GPU launch overhead is only 5us, so I'm not sure this even has a huge impact.

One way to know for sure is to just do a performance run that always uses cudagraphs for benchmarking, so we'll see how that goes (https://github.com/pytorch/pytorch/actions/runs/10184709063).

xuzhao9 replied:
Yeah I agree that if we can get good enough benchmark results without cudagraph, we should not use it.

I am also wondering do you think `torch._inductor.runtime.benchmarking` is good enough for both online and offline benchmarking? Or do you think we could to do additional stuff in an offline benchmarker, which has relaxed constraints on benchmarking time to enable getting higher quality of benchmarking results?

I have a question regarding to the term "GPU launch overhead", I guess it is different from "GPU kernel launch overhead"? I thought the GPU kernel launch overhead is already excluded even without CUDA graph, since we are measuring the

nmacchioni replied:
> I am also wondering do you think torch._inductor.runtime.benchmarking is good enough for both online and offline benchmarking?

If it works online it should work offline

> Or do you think we could to do additional stuff in an offline benchmarker, which has relaxed constraints on benchmarking time to enable getting higher quality of benchmarking results?

Maybe, I'd have to think about it. We would have stricter control over the state of the device we are benchmarking on (i.e. for GPU we could pin the frequency however we wanted), and we'd definitely be able to benchmark for a longer duration. Whether either of these would actually make a difference in E2E performance is anyone's guess. The real winner of offline benchmarking imo would be saving compile time by not having to materialize every template, but at that point we're really just talking about offlining autotuning

nmacchioni replied:
> I have a question regarding to the term "GPU launch overhead", I guess it is different from "GPU kernel launch overhead"? I thought the GPU kernel launch overhead is already excluded even without CUDA graph, since we are measuring the elapsed time between 2 CUDA events?

I think we're talking about the same thing. As for what the cuda events actually record, I believe it does include the kernel launch overhead. My understanding is that when you hit record on our start event it records that instant, then you launch and your kernel gets queued up (hence we incur the launch overhead), it runs, and then we hit record again on our end event.

xuzhao9 replied:
I see, so the "cpu launch overhead" and "gpu kernel launch overhead" are two different metrics. CUDA Graph can eliminate both, while Non-CUDA-Graph can eliminate the first but not the latter (with torch.cuda.Event).

nmacchioni replied:
> I see, so the "cpu launch overhead" and "gpu kernel launch overhead" are two different metrics. CUDA Graph can eliminate both, while Non-CUDA-Graph can eliminate the first but not the latter (with torch.cuda.Event).

Yes, CPU launch overhead is much more variable (for example, consider the case where our kernel needs to allocate an output tensor at runtime) and GPU kernel launch overhead is pretty much static. Benchmarking with cudagraphs eliminates the GPU kernel launch overhead; as for the CPU launch overhead I'm not sure, it should definitely reduce the total overhead but we still need to eat some of it in between recording our start event and the graph replaying. We can completely eliminate CPU launch overhead when not using cudagraphs to benchmark, but indeed we then have to take the GPU kernel launch overhead.

nmacchioni replied:
Seems like `do_bench_cudagraphs` is actually worse for performance (and obviously compile time). My best guess is that we really hurt benchmark validity from skipping the cache clear in between executions of the kernel within the graph, and probably also from the lack of a stable warmup cycle before the graph gets replayed.
![Uploading Screenshot 2024-08-01 at 2.15.28 AM.png…]()

xuzhao9 commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+        # we may as well use the estimation loop signal as well
+        return min(estimated_timing_ms, timing_ms)
+
+    def original_do_bench_gpu(
```
Is there a reason we are copying the code from `triton.testing.do_bench` instead of directly calling the function?

nmacchioni replied:
I just prefer it together (also we don't have to worry about lazily importing triton). Also, the direct copies should be short lived and will be removed post-rollout.

nmacchioni commented:
Custom Triton Pin -> Default Triton Pin
<img width="1711" alt="Screenshot 2024-07-31 at 11 18 14 AM" src="https://github.com/user-attachments/assets/375248ae-8c0e-4318-805f-6a86315bff71">

Seems like no change, I must have originally grouped these changes with some other win and incorrectly attributed them. We can proceed without worrying about upstreaming to Triton. cc @bertmaher @eellison

nmacchioni committed:
temp enable logging by default

nmacchioni committed:
Update test_benchmarking.py

nmacchioni committed:
Update triton_heuristics.py

nmacchioni committed:
fix queue limit off by 1

nmacchioni committed:
fix ut

nmacchioni committed:
Revert "temp enable logging by default"

This reverts commit a0640ddfe579d3b01ec09448f90d38ce948efa52.

nmacchioni committed:
nit

nmacchioni commented:
Latest perf run results:
<img width="1708" alt="Screenshot 2024-08-01 at 1 10 58 PM" src="https://github.com/user-attachments/assets/9444117a-e53b-4f78-a6c5-5fcbf0ddcbff">
<img width="1714" alt="Screenshot 2024-08-01 at 2 15 28 AM" src="https://github.com/user-attachments/assets/0a290afd-c522-42e2-8f1d-ec63f01e60bb">

nmacchioni committed:
temp disable pruning

nmacchioni committed:
temp disable ranking

nmacchioni committed:
temp disable lazy benchmarking

nmacchioni committed:
temp enable fallback

nmacchioni committed:
Revert "temp disable pruning"

This reverts commit 24f7dae73a04498d543df421114d8acb3af040eb.

nmacchioni committed:
Revert "temp disable ranking"

This reverts commit 7824baa29d30e5fc53b28270fefddb2b62e02597.

nmacchioni committed:
Revert "temp disable lazy benchmarking"

This reverts commit e9310bc4447ec34673e2e1803155878129b57c25.

nmacchioni committed:
Revert "temp enable fallback"

This reverts commit 465817976d0831592a32dc2a4dfe5990fef49b0f.

nmacchioni commented:
Doing some perf runs enabling things one at a time:

main -> immediate fallback to original benchmarking (should be same)
<img width="858" alt="Screenshot 2024-08-02 at 12 52 25 AM" src="https://github.com/user-attachments/assets/6ea7d20f-fc4e-433e-a9bf-0a5b1529ce50">

-> enable new benchmarking (non-lazy)
<img width="846" alt="Screenshot 2024-08-02 at 12 52 53 AM" src="https://github.com/user-attachments/assets/f27c7b91-9e62-4091-9716-ce2e1c47760b">

-> enable lazy benchmarking
<img width="845" alt="Screenshot 2024-08-02 at 12 53 15 AM" src="https://github.com/user-attachments/assets/9ec10a59-99a2-4f7f-816d-bddef2ea2676">

-> enable ranking
<img width="838" alt="Screenshot 2024-08-02 at 12 54 01 AM" src="https://github.com/user-attachments/assets/19a5b56e-2a57-4baf-8ed7-da546dbcf9f5">

-> enable pruning
<img width="833" alt="Screenshot 2024-08-02 at 12 54 20 AM" src="https://github.com/user-attachments/assets/210a1a6a-bc37-42e0-95b9-3901cc2b113e">

nmacchioni committed:
fix dropped kwargs

nmacchioni committed:
temp enable fallback

nmacchioni committed:
Revert "temp enable fallback"

This reverts commit 7b0c364be5c52cffd304c122c8a32a50e680329d.

nmacchioni committed:
temp disable pruning

nmacchioni committed:
Revert "temp disable pruning"

This reverts commit e4959da6108ec64e0b0189979236163d7e336564.

nmacchioni commented:
Rerunning fallback perf run with some dropped kwargs fixed, and doing a wider perf run of disabled pruning to see if slowdown is noise

nmacchioni commented:
> Rerunning fallback perf run with some dropped kwargs fixed, and doing a wider perf run of disabled pruning to see if slowdown is noise

Reran fallback, after fixing dropped kwargs, comparing against main (training + inference):
<img width="844" alt="Screenshot 2024-08-03 at 7 32 05 PM" src="https://github.com/user-attachments/assets/99112aac-e271-44ba-a2be-213fddd46270">
<img width="839" alt="Screenshot 2024-08-03 at 7 32 18 PM" src="https://github.com/user-attachments/assets/2f0de391-4c8d-46ee-b912-3ac6f92bdd2e">

Reran enabling pruning, wider benchmark (training + inference):
<img width="848" alt="Screenshot 2024-08-03 at 7 33 16 PM" src="https://github.com/user-attachments/assets/29042cb8-a74e-4bc5-a9e1-966aa18b985a">
<img width="850" alt="Screenshot 2024-08-03 at 7 33 34 PM" src="https://github.com/user-attachments/assets/4c71c26a-4d82-4339-87a6-dffb503cc23f">

Everything looks as expected now. cc @eellison

bertmaher commented:
lgtm but please get a review from a code owner before landing

eellison commented:
To respond some of the comments before full review:

- fallback is slower than main, which shouldn't be the case since we're just slightly wrapping the original do_bench_gpu. I think we're probably dropping some warmup=x or rep=y accidentally and hitting the defaults, so I'll look into this.

Why would this be ? ideally, running benchmarking with this same arguments as we did previously would give the same results. This is also at least for now the api folks would be used to using for new invocations. 

Nice speeds ups on the new benchmarking. It looks like lazy benchmarking does not improve the results - could we improve from this PR (which is already quite large) if adds complexity but does not improve performance ? Similarly, does pruning improve compilation speed ? if not, I would say the same thing.

nmacchioni commented:
> > fallback is slower than main, which shouldn't be the case since we're just slightly wrapping the original do_bench_gpu. I think we're probably dropping some warmup=x or rep=y accidentally and hitting the defaults, so I'll look into this.
> 
> Why would this be ? ideally, running benchmarking with this same arguments as we did previously would give the same results. This is also at least for now the api folks would be used to using for new invocations.

The initial draft of the PR didn't have fallbacks, and didn't maintain consistency in APIs, so initially I just deleted all the do_bench-specific arguments from the callsites. After we discussed this and decided to add in the fallbacks and maintain API consistency, I went back and un-did the changes. In a couple places I missed re-adding the kwargs which was causing the slowdown. This has been fixed, i.e. I just re-added the old kwargs back in.

> Nice speeds ups on the new benchmarking. It looks like lazy benchmarking does not improve the results - could we improve from this PR (which is already quite large) if adds complexity but does not improve performance ?

We shouldn't really see speedups from lazy benchmarking itself, that is not its design. Since it more or less still just does the same benchmarking but delayed. What lazy benchmarking enables us to do is get a group of related benchmarks and run them all at once using the grouped benchmarking, which still doesn't speed things up noticeably. All of this is just getting us to the point where we can do ranking and pruning, which requires that we run multiple benchmarks at once.

> Similarly, does pruning improve compilation speed ? if not, I would say the same thing.

Pruning shouldn't be a factor for default PT2 (which is what we ran to compare turning everything on, to save time) since it is only used in the GEMM autotuning stage (where it 100% does improve overhead)

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+        # we really shouldn't see any significant slowdowns
+        torch.cuda._sleep(
+            int(
+                (self.cpu_launch_overhead_ms_per_event_record * 2500)
```
use variable names for constants

nmacchioni replied:
dne

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+        properties = torch.cuda.get_device_properties(device)
+        return properties.L2_cache_size
+
+    @cached_property
```
huh, i hadnt seen cached_property before...

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+        for _ in range(100):
+            torch.cuda.Event(enable_timing=True).record()
+        torch.cuda.synchronize()
+        return ((perf_counter() - start_time_s) * 1000) / 100
```
use variable names for repeated constants

nmacchioni replied:
done

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+        return self.cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear[1]
+
+    @cached_property
+    def cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear(
```
is there something less verbose we could name this ?

nmacchioni replied:
I couldn't think of anything shorter that would accurately describe the function

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+    @cached_property
+    def cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear(
+        self: Self,
+    ) -> Tuple[float, float]:
```
nit, add docstrings on methods for more complicated ones

nmacchioni replied:
done

eellison reviewed (commented):
partial review

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+    config_val = getattr(benchmarking_config, config_name)
+
+    @lru_cache(None)
+    def is_jk_enabled(name: str) -> bool:
```
yeah in the sense that if the jk changes in the middle of execution we won't record the change. I think I'm fine with this, we save some time checking the jk only once and I think the subset of jobs that would benefit from being disabled midway through is fairly low; if there is some issue with benchmarking I assume most jobs would just crash or have garbage output if even one benchmarking was already finished, so if we've already done one benchmarking with the jk enabled then it got disabled for the rest of the remaining benchmarks I don't think we'd save the job.

nmacchioni replied:
removed lru_cache anyways, checked and jk should be really fast

nmacchioni committed:
defaults in config.py

nmacchioni committed:
don't cache jk

nmacchioni committed:
remove now unused magic methods

nmacchioni committed:
dedup fallbacks

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+            return self.timing_ms - other
+        return LazyBenchmark(lambda: self.timing_ms - other)
+
+    def __rsub__(self: Self, other: Any) -> Any:
```
removing these functions, they are unused and just noisy at this time

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+        return self.timing_ms >= other
+
+    def __add__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return self.timing_ms + other
```
self is never not a LazyBenchmark

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+)
+
+
+def maybe_fallback_to_original_benchmarking(
+    original_fn_name: str,
+) -> Callable[..., Any]:
+    def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(fn)
+        def wrapper(self: Benchmarker, *args: Any, **kwargs: Any) -> Any:
+            if should_fallback_to_original_benchmarking():
+                log.debug(
+                    "Falling back to original benchmarking function %s from %s.",
+                    original_fn_name,
+                    fn.__name__,
+                )
+                counters["inductor"][
+                    "benchmarking_fallback_to_original_benchmarking"
+                ] += 1
+                return getattr(self, original_fn_name)(*args, **kwargs)
+            return fn(self, *args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def maybe_fallback_to_non_lazy_benchmarking(
+    non_lazy_fn_name: str,
+) -> Callable[..., Any]:
+    def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(fn)
```
done

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+    "fallback_to_original_benchmarking",
+    False,
+    False,
+    True,
```
done

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+            return self.timing_ms * other
+        return LazyBenchmark(lambda: self.timing_ms * other)
+
+    def __rmul__(self: Self, other: Any) -> Any:
```
never, removed these as I only implemented them to cover future cases but we can just implement as we go if needed

nmacchioni committed:
no magic numbers

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,1001 @@
+                / self.gpu_time_ms_per_gpu_clock_cycle
+            )
+        )
+        for idx in range(2500):
+            start_time_s = perf_counter()
+            torch.cuda.Event(enable_timing=True).record()
+            elapsed_time_ms = (perf_counter() - start_time_s) * 1000
+            # recording an event is near instantaneous, unless we have hit
```
there is no official spec because it is liable to change at any time

nmacchioni committed:
docstrings

nmacchioni committed:
remove now unused tests

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+        estimated_timing_ms = self.get_min_timing_ms(event_pairs)
+
+        # we don't want to benchmark for longer than max_benchmark_duration milliseconds,
+        # so we reduce benchmark_iters to fit within this constraint if necessary
+        benchmark_iters = self.reduce_benchmark_iters_to_max_benchmark_duration_ms(
+            memory_warmup_iters,
+            benchmark_iters,
+            cpu_launch_overhead_ms_per_iter,
+            estimated_timing_ms,
+            1,
+            max_benchmark_duration_ms,
+        )
+        # the self.gpu_queue_limit'th event queued to the GPU will execute synchronously
+        # on the CPU; this is bad since we try to overlap the CPU launch period with a
+        # GPU sleep to avoid the CPU launch overhead in our GPU event timings. what ends
+        # up happening in this case is that the self.gpu_queue_limit'th event will wait
+        # until the queue shrinks (i.e. after the GPU sleep ends) before queueing itself.
+        # this has two issues, first we will waste a lot of time needlesly as the CPU
+        # waits for the GPU to finish its sleep, and second we will no longer have a
+        # proper overlap between the CPU launch overhead and the GPU sleep which can
+        # result in incorrect timing values for particularly small kernels as the CPU
+        # will eventually catch up to the GPU and we will end up in lock-step execution.
+        # we can prevent this by benchmarking in blocks. we calculate the number of blocks
+        # required by assuming that each benchmark iteration spawns approximately 5 events:
+        # 1 for the buffer zeroing, 1 for the callable, 2 for the event records, and 1
+        # extra allocated for any sub-kernels the callable may launch of which the most
+        # common case is allocating an output Tensor. we then evenly split the total
+        # number of benchmark iterations across all blocks.
```
Similar to some of the other questions - did we benchmark how helpful this was actually was ? From the comments in your draft to upstream triton, I remember them commenting how stable cuda kernels were, for the most part.

nmacchioni replied:
sleeps brought us on par performance-wise with the original do_bench implementation, by removing the chance of CPU-GPU lock stepping

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+        # number of benchmark iterations across all blocks.
+        benchmark_blocks = ((benchmark_iters * 5) // (self.gpu_queue_limit - 1)) + 1
+        benchmark_iters_per_block = benchmark_iters // benchmark_blocks
+        memory_warmup_iters_per_block = memory_warmup_iters // benchmark_blocks
```
cc @ipiszy (who added do_bench_profiling) and @int3 (who I believed looked into something simliar) for comment on this section

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+        del buffer
+
+        # we may as well use the estimation loop signal as well
+        return min(estimated_timing_ms, timing_ms)
```
I wold also call out in the docstring for this function that it will return the minimum time which is a divergence from the existing triton benchmarking. 

also, cc, @int3, @bertmaher  if you have any thoughts. I know there have been some discussions previously about benchmarking with different clock rates

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+                start_event.record()
+                _callable()
+                end_event.record()
+        cpu_launch_overhead_ms_per_iter = (
```
I dont that its a valid assumption that all of the kernels will have a proportional cpu launch overhead

nmacchioni replied:
But they should be *close enough*, the sleep duration doesn't have to be entirely perfect we just want to avoid sleeping for 20x choices when 15x were pruned

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+            interleaved_event_pairs
+        )
+
+        if ranking_key is not None:
+            if should_enable_early_ranking():
+                log.debug(
```
Just for my own understanding - enable_early_ranking is the default today ? And we might want to fallback to this if there is an error with the new logic ? If so, `should_enable` is not api i would expect.

nmacchioni replied:
`enable_early_ranking` defaults to True in OSS, False internal. Would you expect `should_disable`?

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+        pruning_factor: float = 1.1,
+        pruning_limit: int = 5,
+        **kwargs: Any,
+    ) -> List[float]:
```
can we get docstrings for all of the methods ?

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+            len(callables_to_benchmark),
+            max_benchmark_duration_ms,
+        )
+        benchmark_blocks = (
+            (benchmark_iters * 5 * len(callables_to_benchmark))
+            // (self.gpu_queue_limit - 1)
+        ) + 1
+        benchmark_iters_per_block = benchmark_iters // benchmark_blocks
+        memory_warmup_iters_per_block = memory_warmup_iters // benchmark_blocks
+        required_gpu_sleep_cycles_per_block = self.get_required_gpu_sleep_cycles(
+            memory_warmup_iters_per_block,
+            benchmark_iters_per_block,
+            cpu_launch_overhead_ms_per_iter,
+        )
```
The reason for `5` is commented above but not clear here. Could we factor out some of this ?

eellison commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+        interleaved_event_pairs = self.get_interleaved_event_pairs(
+            len(callables_to_benchmark), benchmark_iters
+        )
+        for block_idx in range(benchmark_blocks):
+            block_start = block_idx * benchmark_iters_per_block
+            if block_idx == (benchmark_blocks - 1):
+                block_end = benchmark_iters
+            else:
+                block_end = (block_idx + 1) * benchmark_iters_per_block
+            torch.cuda._sleep(required_gpu_sleep_cycles_per_block)
```
possible to factor out some of this as well ?

eellison reviewed (commented):
partial review. would be great to have some input from folks who have more closely looked at benchmarking like @int3 and @ipiszy

nmacchioni commented on torch/_inductor/runtime/benchmarking.py:
```diff
@@ -0,0 +1,944 @@
+
+        event_pairs = self.get_event_pairs(estimation_iters)
+        start_time_s = perf_counter()
+        for start_event, end_event in event_pairs:
```
> I think we should expose whether to remove cpu overhead as an argument. In some cases, the overhead is part of the benchmark.

Is this for the fusion case? The current `do_bench` implementation should (mostly) remove cpu overhead, so if we wanted a way to accurately measure cpu overhead + gpu time that would probably have to be a followup.

> Additionally - we're doing 5 iters here.. I know with torch.compile & cudagraphs we only hit the fast path in the 3rd iteration. So we would be disproportionately aiding the cudagrpah backend here. Granted, thats uncommon case, I wouldnt really expect us to hit in this benchmarking.. but i think is worth exposing as an arg and documenting

It is exposed by `estimation_iters`, and is the same exact default as before.

nmacchioni commented:
@eellison Let me just go ahead and break this PR up, otherwise I feel we'll get stuck in a feedback loop haha
